### PR TITLE
Bump fpdf2 to 2.5.*

### LINKED ISSUE #7792: stubs for fpdf2 have errors and are out-of-date
I am encountering a number of issues using pyright with the fpdf2 library, specifically with the fpdf.pyi file.

1) There are 4 type aliases to Literal strings: `_xxx: TypeAlias = Literal["", ...]`
These incorrectly throw errors when a string variable (as opposed to string Literal) is used. Maybe they should just be `str`??

2) The` cell()` and `multicell()` methods now accept arguments named `new_x` and `new_y`. I suspect this package has not been updated in a while to support new/changed functionality in fpdf2. The current version is 2.5.2 and the stubs indicate they are for 2.4.*.

I guess a more general question would be "How do I get 3rd party packages updated?"

### PR DESCRIPTION
**Pull request by hauntsaninja**
Release: https://pypi.org/project/fpdf2/2.5.5/
Homepage: https://pyfpdf.github.io/fpdf2/

Mostly fixes #7792.

### CODE REVIEW

File: stubs/fpdf2/fpdf/actions.pyi [hunk 1]
```diff
@@ -1,26 +1,29 @@
+from _typeshed import Incomplete
 from abc import ABC
-from typing import Any
+
+from .syntax import PDFObject
 
 class Action(ABC):
-    def dict_as_string(self) -> None: ...
+    def __init__(self, next_action: PDFObject | str | None = ...): ...
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/actions.pyi:7]:
```suggestion
    next: PDFObject | str | None
    def __init__(self, next_action: PDFObject | str | None = ...): ...
```

File: stubs/fpdf2/fpdf/actions.pyi [hunk 2]
```diff
@@ -1,26 +1,29 @@
...
+    def dict_as_string(self, key_values: dict[str, Incomplete] | None = ...) -> None: ...
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/actions.pyi:8]: This returns whatever a call to `create_dictionary_string` returns: https://github.com/PyFPDF/fpdf2/blob/be3ad07b7ae9c7d39866aa130de177b43fc3f073/fpdf/actions.py#L23

And `create_dictionary_string` always returns `str`: https://github.com/PyFPDF/fpdf2/blob/be3ad07b7ae9c7d39866aa130de177b43fc3f073/fpdf/syntax.py#L99

```suggestion
    def dict_as_string(self, key_values: dict[str, Incomplete] | None = ...) -> str: ...
```

File: stubs/fpdf2/fpdf/actions.pyi [hunk 3]
```diff
@@ -1,26 +1,29 @@
...
 
 class NamedAction(Action):
-    action_name: Any
-    def __init__(self, action_name) -> None: ...
+    action_name: Incomplete
+    def __init__(self, action_name, next_action: PDFObject | str | None = ...) -> None: ...
     def dict_as_string(self): ...
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/actions.pyi:13]:
```suggestion
    def dict_as_string(self) -> str: ...
```

File: stubs/fpdf2/fpdf/actions.pyi [hunk 4]
```diff
@@ -1,26 +1,29 @@
...
 
 class GoToAction(Action):
-    dest: Any
-    def __init__(self, dest) -> None: ...
+    dest: Incomplete
+    def __init__(self, dest, next_action: PDFObject | str | None = ...) -> None: ...
     def dict_as_string(self): ...
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/actions.pyi:18]:
```suggestion
    def dict_as_string(self) -> str: ...
```

File: stubs/fpdf2/fpdf/actions.pyi [hunk 5]
```diff
@@ -1,26 +1,29 @@
...
 
 class GoToRemoteAction(Action):
-    file: Any
-    dest: Any
-    def __init__(self, file, dest) -> None: ...
+    file: Incomplete
+    dest: Incomplete
+    def __init__(self, file, dest, next_action: PDFObject | str | None = ...) -> None: ...
     def dict_as_string(self): ...
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/actions.pyi:24]:
```suggestion
    def dict_as_string(self) -> str: ...
```

File: stubs/fpdf2/fpdf/actions.pyi [hunk 6]
```diff
@@ -1,26 +1,29 @@
...
 
 class LaunchAction(Action):
-    file: Any
-    def __init__(self, file) -> None: ...
+    file: Incomplete
+    def __init__(self, file, next_action: PDFObject | str | None = ...) -> None: ...
     def dict_as_string(self): ...
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/actions.pyi:29]:
```suggestion
    def dict_as_string(self) -> str: ...
```

File: stubs/fpdf2/fpdf/enums.pyi [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    L: str
+    R: str
+    J: str
+
+class RenderStyle(CoerciveEnum):
+    D: str
+    F: str
+    DF: str
+    @property
+    def operator(self): ...
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/enums.pyi:32]:
```suggestion
    def operator(self) -> str: ...
```

File: stubs/fpdf2/fpdf/enums.pyi [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    @property
+    def is_draw(self): ...
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/enums.pyi:34]:
```suggestion
    def is_draw(self) -> bool: ...
```

File: stubs/fpdf2/fpdf/enums.pyi [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    @property
+    def is_fill(self): ...
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/enums.pyi:36]:
```suggestion
    def is_fill(self) -> bool: ...
```

File: stubs/fpdf2/fpdf/enums.pyi [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    @classmethod
+    def coerce(cls, value): ...
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/enums.pyi:38]: This looks to me like it has the same signature as the superclass, I think we can delete it here

File: stubs/fpdf2/fpdf/fpdf.pyi [hunk 11]
```diff
@@ -33,10 +34,19 @@ class Annotation(NamedTuple):
...
     height: int
+    flags: tuple[AnnotationFlag, ...] = ...
     contents: str | None = ...
     link: str | int | None = ...
     alt_text: str | None = ...
     action: Action | None = ...
+    color: int | None = ...
+    modification_time: datetime.datetime | None = ...
+    title: str | None = ...
+    quad_points: Sequence[int] | None = ...
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/fpdf.pyi:45]: This is typed as a `tuple` in the source code, any reason to use `Sequence` here?

File: stubs/fpdf2/fpdf/fpdf.pyi [hunk 12]
```diff
@@ -128,9 +137,11 @@ class FPDF:
         orientation: _Orientation = ...,
         unit: _Unit | float = ...,
         format: _Format | tuple[float, float] = ...,
-        font_cache_dir: bool = ...,
+        font_cache_dir: bool | Literal["DEPRECATED"] = ...,
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/fpdf.pyi:140]: The docstring says "path or str". But it's also deprecated and ignored at runtime, so it doesn't really matter much.

https://github.com/PyFPDF/fpdf2/blob/be3ad07b7ae9c7d39866aa130de177b43fc3f073/fpdf/fpdf.py#L329

**Review by AlexWaygood — CHANGES_REQUESTED** [review]

File: stubs/fpdf2/fpdf/actions.pyi (hunk 1 above)
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/actions.pyi:7]:
```suggestion
    def __init__(self, next_action: PDFObject | str | None = ...) -> None : ...
```

File: stubs/fpdf2/fpdf/enums.pyi (hunk 10 above)
Comment (hauntsaninja) [inline stubs/fpdf2/fpdf/enums.pyi:38]:
```suggestion
```

File: stubs/fpdf2/fpdf/fpdf.pyi (hunk 11 above)
Comment (hauntsaninja) [reply stubs/fpdf2/fpdf/fpdf.pyi:45]: Yeah, I saw some place in the source where they make a list and pass it around

File: stubs/fpdf2/fpdf/fpdf.pyi (hunk 12 above)
Comment (hauntsaninja) [inline stubs/fpdf2/fpdf/fpdf.pyi:140]:
```suggestion
        font_cache_dir: str | Literal["DEPRECATED"] = ...,
```

File: stubs/fpdf2/fpdf/actions.pyi [hunk 13]
```diff
@@ -1,26 +1,29 @@
+from _typeshed import Incomplete
 from abc import ABC
-from typing import Any
+
+from .syntax import PDFObject
 
 class Action(ABC):
-    def dict_as_string(self) -> None: ...
+    def __init__(self, next_action: PDFObject | str | None = ...) -> None : ...
```
Comment (hauntsaninja) [inline stubs/fpdf2/fpdf/actions.pyi]:
```suggestion
    next: PDFObject | str | None
    def __init__(self, next_action: PDFObject | str | None = ...) -> None : ...
```

File: stubs/fpdf2/fpdf/actions.pyi [hunk 14]
```diff
@@ -1,26 +1,29 @@
+from _typeshed import Incomplete
 from abc import ABC
-from typing import Any
+
+from .syntax import PDFObject
 
 class Action(ABC):
-    def dict_as_string(self) -> None: ...
+    def __init__(self, next_action: PDFObject | str | None = ...) -> None: ...
```
Comment (AlexWaygood) [inline stubs/fpdf2/fpdf/actions.pyi:7]:
```suggestion
    next: PDFObject | str | None
    def __init__(self, next_action: PDFObject | str | None = ...) -> None: ...
```

**Review by AlexWaygood — APPROVED** [review]
Just one last thing, otherwise I'm happy!

File: stubs/fpdf2/fpdf/actions.pyi (hunk 14 above)
Comment (hauntsaninja) [reply stubs/fpdf2/fpdf/actions.pyi:7]: Ahh, sorry, I thought I'd got that, but I think Github didn't apply the suggestion because of a merge conflict.

**Comment by AlexWaygood** [comment]
Ah, I see GitHub Actions is down: https://www.githubstatus.com

**Pull request merged**